### PR TITLE
BUG: .rename* not treating Series as mapping

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pydata/pandas/issues/12623

I added com.is_dict_like in https://github.com/pydata/pandas/pull/11980
and failed to use it for the `rename` method. Using that now and did
some refactoring while I was in there. Added more tests for rename.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
I might remove or rename `core._is_scalar_or_list_like`. It doesn't seem like a well-defined or clear check...

**Comment by jreback** [comment]
@TomAugspurger btw, make sure you are rebased on current master. ``dateutil=2.5.0`` flowed thru and broke some tests and I just fixed in (maybe  2 hrs ago)

File: pandas/core/series.py [hunk 1]
```diff
@@ -2331,11 +2330,7 @@ def align(self, other, join='outer', axis=None, level=None, copy=True,
 
     @Appender(generic._shared_docs['rename'] % _shared_doc_kwargs)
     def rename(self, index=None, **kwargs):
-        is_scalar_or_list = (
-            (not com.is_sequence(index) and not callable(index)) or
-            (com.is_list_like(index) and not isinstance(index, MutableMapping))
-        )
-        if is_scalar_or_list:
```
Comment (jreback) [inline pandas/core/series.py]: yeah I think ``lib.isscalar(...) or com.is_list_like(...)`` is more clear

File: pandas/tests/series/test_alter_axes.py [hunk 2]
```diff
@@ -55,6 +55,13 @@ def test_rename(self):
         renamed = renamer.rename({})
         self.assertEqual(renamed.index.name, renamer.index.name)
 
+    def test_rename_by_series(self):
+        s = Series(range(5), name='foo')
+        renamer = Series({1: 10, 2: 20})
```
Comment (jreback) [inline pandas/tests/series/test_alter_axes.py]: maybe tests both dict and Series.

```
d = {1: 10, 2: 20}
for value in [ d, Series(d) ]:
     ....
```
for

File: pandas/core/series.py (hunk 1 above)
Comment (TomAugspurger) [inline pandas/core/series.py]: `com.is_list_like(dict)` is True, and  I need to treat dicts differently than lists.

But if we're going to be raising on `Series.rename(list)` anway, I won't have to worry about lists at all...

File: pandas/tests/series/test_alter_axes.py (hunk 2 above)
Comment (TomAugspurger) [inline pandas/tests/series/test_alter_axes.py]: These should all be in `test_generic` now. Initially I didn't understand how `test_generic` was setup, so I added the test here. I *think* these tests are all redundant, and could remove them this one and `test_rename` above, but I always worry about deleting tests :)

Comment (jreback) [inline pandas/tests/series/test_alter_axes.py]: cool np. yeah. hard to stop the propogation of tests :)

**Comment by TomAugspurger** [comment]
Ok, the tests for the different possible arguments are [here](https://github.com/pydata/pandas/pull/12626/files#diff-f042fb8fba5656b3854ad92563ed6480R93). I simplified the stuff with `is_scalar_or_mapping` to just do the check in the code. Should be clearer now, but a bit less DRY.

**Comment by jreback** [comment]
thanks @TomAugspurger

**Pull request closed without merging**